### DISCUSSION TITLE
how to migrate properties that both contain observer and computed to lit

### DISCUSSION THREAD

**Discussion by paladox** [lit/lit#2778] [Q&A] [answered]
I'm wondering if someone can tell me how i can migrate https://github.com/GerritCodeReview/gerrit/blob/master/polygerrit-ui/app/elements/change/gr-change-actions/gr-change-actions.ts#L461 to L467 to lit please?

It does both computed and a observer so i'm not sure which way is best?

**Comment by kevinpschaaf** [comment, marked as answer]
Looks like the component has been migrated, so it looks like you found the equivalent patterns in Lit?

For others who may come across this, here are some general rules of thumb:

* computed properties
  * If only used ephemerally in the template, compute as local variables in `render()` and use in template before returning
  * If they need to be stored persistently or are expensive to compute, do so in `willUpdate(changedProperties)`
  * Use the `changedProperties.has('xyz')` to compute only when dependencies change, to avoid expensive re-computation every update
* observers
  * If the observer needs to act directly on DOM based on property changes, use `updated()`
  * Otherwise, use `willUpdate()`
  * In either case, use `changedProperties.has('xyz')` to know when a property has changed
* path-based observers (`observers: ['fooBarChanged(foo.bar)', 'fooChanged(foo.*)'])`)
  * This was a feature very specific to Polymer's data system that has no equivalent in Lit.
  * We recommend using immutable patterns as a more interoperable way to communicate deep property changes; see [Mutating Object and Array Properties](https://lit.dev/docs/components/properties/#mutating-properties) for more info.

@arthurevans is currently working on a "Lit for Polymer Users" guide that will go up on the lit.dev site soon, which should be a more permanent resource for these type of question.
